sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainer.java

```diff
@@ -0,0 +1,201 @@
+ *   {@link #getDeltas()} the runner should invoke {@link #commitDeltas} to indicate that those
+ *   deltas have been incorporated.
+ *   </li>
+ * </ol>
+ */
+@Experimental(Kind.METRICS)
+public class MetricsContainer {
+
+  private static final ThreadLocal<MetricsContainer> CONTAINER_FOR_THREAD =
+      new ThreadLocal<MetricsContainer>();
+  private final AtomicBoolean committedFirstDelta = new AtomicBoolean(false);
+
+  private final String stepName;
+
+  private MetricsMap<MetricName, CounterCell> counters =
+      new MetricsMap<MetricName, CounterCell>() {
+        @Override
+        protected CounterCell createInstance() {
+          return new CounterCell();
+        }
+      };
+
+  private MetricsMap<MetricName, DistributionCell> distributions =
+      new MetricsMap<MetricName, DistributionCell>() {
+        @Override
+        protected DistributionCell createInstance() {
+          return new DistributionCell();
+        }
+      };
+
+  /**
+   * Create a new {@link MetricsContainer} associated with the given {@code stepName}.
+   */
+  public MetricsContainer(String stepName) {
+    this.stepName = stepName;
+  }
+
+  /**
+   * Set the {@link MetricsContainer} for the current thread.
+   */
+  public static void setMetricsContainer(MetricsContainer container) {
+    CONTAINER_FOR_THREAD.set(container);
+  }
+
+  /**
+   * Clear the {@link MetricsContainer} for the current thread.
+   */
+  public static void unsetMetricsContainer() {
+    CONTAINER_FOR_THREAD.remove();
+  }
+
+  /**
+   * Return the {@link MetricsContainer} for the current thread.
+   */
+  public static MetricsContainer getCurrentContainer() {
+    MetricsContainer container = CONTAINER_FOR_THREAD.get();
+    if (container == null) {
+      throw new IllegalStateException("Must call setMetricsContainer before reporting metrics.");
+    }
+    return container;
+  }
+
+  /**
+   * Return the {@link CounterCell} that should be used for implementing the given
+   * {@code metricName} in this container.
+   */
+  public CounterCell getOrCreateCounter(MetricName metricName) {
+    return counters.getOrCreate(metricName);
+  }
+
+  public DistributionCell getOrCreateDistribution(MetricName metricName) {
+    return distributions.getOrCreate(metricName);
+  }
+
+  private <UpdateT, CellT extends MetricCell<UpdateT>>
+  ImmutableList<MetricUpdate<UpdateT>> extractUpdates(
+      MetricsMap<MetricName, CellT> cells,
+      boolean includeZero) {
+    ImmutableList.Builder<MetricUpdate<UpdateT>> updates = ImmutableList.builder();
+    for (Map.Entry<MetricName, CellT> cell : cells.entries()) {
+      UpdateT update = cell.getValue().getDeltaUpdate(includeZero);
+      if (update != null) {
+        updates.add(MetricUpdate.create(MetricKey.create(stepName, cell.getKey()), update));
+      }
+    }
+    return updates.build();
+  }
+
+  /**
+   * Return the delta updates for all metrics in this container.
+   */
```

Comment:
This logic has changed significantly based on other comments -- suspect this is defunct.